crates/sui-framework/sources/test/random.move

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[test_only]
+module sui::random {
+    use std::hash;
+    use std::vector;
+    use sui::bcs;
+
+    // Internally, the pseudorandom generator uses a hash chain over Sha3-256
+    // which has an output length of 32 bytes.
+    const DIGEST_LENGTH: u64 = 32;
+
+    /// This represents a seeded pseudorandom generator. Note that the generated
+    /// values are not safe to use for cryptographic purposes.
+    struct Random has store, drop {
+        state: vector<u8>,
+    }
+
+    /// Update the state of the generator and return a vector holding `DIGEST_LENGTH`
+    /// random bytes.
+    fun next_digest(random: &mut Random): vector<u8> {
+        random.state = hash::sha3_256(random.state);
+        random.state
+    }
+
+    /// Create a new pseudorandom generator with the given seed.
+    public fun new(seed: vector<u8>): Random {
+        Random { state: seed }
+    }
+
+    /// Use the given pseudorandom generator to generate a vector with l random bytes.
+    public fun next_bytes(random: &mut Random, l: u64): vector<u8> {
+        // We need ceil(l / DIGEST_LENGTH) digests to fill the array
+        let quotient = l / DIGEST_LENGTH;
+        let remainder = l - quotient * DIGEST_LENGTH;
+
+        let (i, output) = (0, vector::empty<u8>());
+        while (i < quotient) {
+            vector::append(&mut output, next_digest(random));
+            i = i + 1;
+        };
+
+        // If quotient is not exact, fill the remaining bytes
+        if (remainder > 0) {
+            let (i, digest) = (0, next_digest(random));
+            while (i < remainder) {
+                vector::push_back(&mut output, *vector::borrow(&mut digest, i));
+                i = i + 1;
+            };
+        };
+
+        output
+    }
+
+    /// Use the given pseudorandom generator to generate a random `u64` integer.
+    public fun next_u64(random: &mut Random): u64 {
+        let bytes = next_digest(random);
+        bcs::peel_u64(&mut bcs::new(bytes))
+    }
+
+    /// Use the given pseudorandom generator to generate an `u64` integer in the range
+    /// [0, ..., 2^bit_length - 1].
+    fun next_u64_with_bit_length(random: &mut Random, bit_length: u8): u64 {
+        next_u64(random) >> (64 - bit_length)
```

Comment:
Putting the `bit_length()` call in here instead of asking the caller to do it might make it more obvious that `64 - bit_length` can't overflow.